File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        TreeNode* curr = root;
        while(curr){
            if(curr->left == nullptr){
                k--;
                if(k == 0) return curr->val;
                curr = curr->right;
            } else {
                TreeNode* next = curr -> left;
                while(next->right && next->right !=curr) {
                    next = next->right;
                }
                if(next->right) {
                    next ->right = nullptr;
                    k--;
                    if(!k) return curr->val;
                    curr = curr ->right;
                } else {
                    next->right = curr;
                    curr = curr->left;
                }
            }
        }
        return -1;
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);

        while(!q.isEmpty()) {
            TreeNode right = null;

            int n = q.size();

            for(int i = 0;i<n;i++) {
                TreeNode curr = q.poll();
                if(curr != null ){
                    right = curr;
                    q.offer(right.left);
                    q.offer(curr.right);
                }
            }
            if(right != null) 
                ans.add(right.val);
        }
        return ans;
    }
}
